File: python/mlkem/ntt.py
"""Number Theoretic Transform (NTT) operations per FIPS 203.

Implements Algorithms 9-12.
"""

from mlkem.field import Q, mod_q, field_mul, field_sub, field_add


def bit_rev7(n: int) -> int:
    """Reverse the 7 least significant bits of n."""
    result = 0
    for _ in range(7):
        result = (result << 1) | (n & 1)
        n >>= 1
    return result


def _precompute_zetas() -> list[int]:
    """Precompute zeta values: zetas[i] = 17^(bitrev7(i)) mod Q."""
    zetas = []
    for i in range(128):
        zetas.append(pow(17, bit_rev7(i), Q))
    return zetas


ZETAS: list[int] = _precompute_zetas()


def ntt(f: list[int]) -> list[int]:
    """Algorithm 9: Number-Theoretic Transform.

    Input: polynomial f with 256 coefficients in Z_Q.
    Output: NTT representation f_hat.
    """
    f_hat = list(f)
    k = 1
    length = 128
    while length >= 2:
        start = 0
        while start < 256:
            zeta = ZETAS[k]
            k += 1
            for j in range(start, start + length):
                t = field_mul(zeta, f_hat[j + length])
                f_hat[j + length] = field_sub(f_hat[j], t)
                f_hat[j] = field_add(f_hat[j], t)
            start += 2 * length
        length //= 2
    return f_hat


def ntt_inverse(f_hat: list[int]) -> list[int]:
    """Algorithm 10: Inverse NTT.

    Input: NTT representation f_hat.
    Output: polynomial f with 256 coefficients in Z_Q.
    """
    f = list(f_hat)
    k = 127
    length = 2
    while length <= 128:
        start = 0
        while start < 256:
            zeta = ZETAS[k]
            k -= 1
            for j in range(start, start + length):
                t = f[j]
                f[j] = field_add(t, f[j + length])
                f[j + length] = field_mul(zeta, field_sub(f[j + length], t))
            start += 2 * length
        length *= 2
    # Multiply all coefficients by 3303 = 128^{-1} mod Q
    inv128 = 3303
    f = [field_mul(c, inv128) for c in f]
    return f


def base_case_multiply(a0: int, a1: int, b0: int, b1: int, gamma: int) -> tuple[int, int]:
    """Algorithm 12: Base case multiplication.

    Multiply two degree-1 polynomials modulo X^2 - gamma.
    """
    c0 = field_add(field_mul(a0, b0), field_mul(field_mul(a1, b1), gamma))
    c1 = field_add(field_mul(a0, b1), field_mul(a1, b0))
    return c0, c1


def multiply_ntts(f_hat: list[int], g_hat: list[int]) -> list[int]:
    """Algorithm 11: Multiplication in NTT domain.

    Pointwise multiplication of two NTT-domain polynomials.
    """
    h_hat = [0] * 256
    for i in range(64):
        gamma = ZETAS[64 + i]
        z0, z1 = base_case_multiply(
            f_hat[4 * i], f_hat[4 * i + 1],
            g_hat[4 * i], g_hat[4 * i + 1],
            gamma,
        )
        h_hat[4 * i] = z0
        h_hat[4 * i + 1] = z1

        z0, z1 = base_case_multiply(
            f_hat[4 * i + 2], f_hat[4 * i + 3],
            g_hat[4 * i + 2], g_hat[4 * i + 3],
            mod_q(Q - gamma),
        )
        h_hat[4 * i + 2] = z0
        h_hat[4 * i + 3] = z1
    return h_hat

File: php/src/MlKem/Ntt.php
<?php

declare(strict_types=1);

namespace PQC\MlKem;

/**
 * Number Theoretic Transform for ML-KEM (q=3329, zeta=17).
 */
final class Ntt
{
    /**
     * Precomputed zetas: zeta^{BitRev7(i)} mod q for i=0..127.
     * zeta = 17, q = 3329.
     */
    private static ?array $zetas = null;

    public static function getZetas(): array
    {
        if (self::$zetas !== null) {
            return self::$zetas;
        }

        // Precomputed zetas for ML-KEM NTT
        self::$zetas = [
            1, 1729, 2580, 3289, 2642, 630, 1897, 848,
            1062, 1919, 193, 797, 2786, 3260, 569, 1746,
            296, 2447, 1339, 1476, 3046, 56, 2240, 1333,
            1426, 2094, 535, 2882, 2393, 2879, 1974, 821,
            289, 331, 3253, 1756, 1197, 2304, 2277, 2055,
            650, 1977, 2513, 632, 2865, 33, 1320, 1915,
            2319, 1435, 807, 452, 1438, 2868, 1534, 2402,
            2647, 2617, 1481, 648, 2474, 3110, 1227, 910,
            17, 2761, 583, 2649, 1637, 723, 2288, 1100,
            1409, 2662, 3281, 233, 756, 2156, 3015, 3050,
            1703, 1651, 2789, 1789, 1847, 952, 1461, 2687,
            939, 2308, 2437, 2388, 733, 2337, 268, 641,
            1584, 2298, 2037, 3220, 375, 2549, 2090, 1645,
            1063, 319, 2773, 757, 2099, 561, 2466, 2594,
            2804, 1092, 403, 1026, 1143, 2150, 2775, 886,
            1722, 1212, 1874, 1029, 2110, 2935, 885, 2154,
        ];

        return self::$zetas;
    }

    /**
     * Forward NTT: transforms polynomial from normal to NTT domain.
     *
     * @param array $f Polynomial coefficients (256 integers)
     * @return array NTT-domain coefficients
     */
    public static function ntt(array $f): array
    {
        $zetas = self::getZetas();
        $k = 1;
        $len = 128;

        while ($len >= 2) {
            for ($start = 0; $start < 256; $start += 2 * $len) {
                $z = $zetas[$k++];
                for ($j = $start; $j < $start + $len; $j++) {
                    $t = Field::mul($z, $f[$j + $len]);
                    $f[$j + $len] = Field::sub($f[$j], $t);
                    $f[$j] = Field::add($f[$j], $t);
                }
            }
            $len >>= 1;
        }

        return $f;
    }

    /**
     * Inverse NTT: transforms from NTT domain back to normal.
     *
     * @param array $f NTT-domain coefficients
     * @return array Normal polynomial coefficients
     */
    public static function invNtt(array $f): array
    {
        $zetas = self::getZetas();
        $k = 127;
        $len = 2;

        while ($len <= 128) {
            for ($start = 0; $start < 256; $start += 2 * $len) {
                $z = $zetas[$k--];
                for ($j = $start; $j < $start + $len; $j++) {
                    $t = $f[$j];
                    $f[$j] = Field::add($t, $f[$j + $len]);
                    $f[$j + $len] = Field::mul($z, Field::sub($f[$j + $len], $t));
                }
            }
            $len <<= 1;
        }

        // Multiply by 128^{-1} mod q (7-layer NTT for ML-KEM)
        $nInv = Field::inv(128);
        for ($i = 0; $i < 256; $i++) {
            $f[$i] = Field::mul($f[$i], $nInv);
        }

        return $f;
    }

    /**
     * Base case multiplication for two NTT-domain elements.
     * Multiplies (a0 + a1*X) * (b0 + b1*X) mod (X^2 - gamma).
     */
    public static function baseMul(int $a0, int $a1, int $b0, int $b1, int $gamma): array
    {
        $c0 = Field::add(Field::mul($a0, $b0), Field::mul(Field::mul($a1, $b1), $gamma));
        $c1 = Field::add(Field::mul($a0, $b1), Field::mul($a1, $b0));
        return [$c0, $c1];
    }

    /**
     * Multiply two polynomials in NTT domain.
     */
    public static function mulNtt(array $a, array $b): array
    {
        $zetas = self::getZetas();
        $c = array_fill(0, 256, 0);

        for ($i = 0; $i < 64; $i++) {
            $z = $zetas[64 + $i];
            [$c[4*$i], $c[4*$i+1]] = self::baseMul(
                $a[4*$i], $a[4*$i+1], $b[4*$i], $b[4*$i+1], $z
            );
            [$c[4*$i+2], $c[4*$i+3]] = self::baseMul(
                $a[4*$i+2], $a[4*$i+3], $b[4*$i+2], $b[4*$i+3], Field::mod(-$z)
            );
        }

        return $c;
    }

    /**
     * Add two polynomials (coefficient-wise).
     */
    public static function polyAdd(array $a, array $b): array
    {
        $c = [];
        for ($i = 0; $i < 256; $i++) {
            $c[$i] = Field::add($a[$i], $b[$i]);
        }
        return $c;
    }

    /**
     * Subtract two polynomials.
     */
    public static function polySub(array $a, array $b): array
    {
        $c = [];
        for ($i = 0; $i < 256; $i++) {
            $c[$i] = Field::sub($a[$i], $b[$i]);
        }
        return $c;
    }
}
